invoke.go

```diff
@@ -33,37 +37,60 @@ func invokeWithRetry(ctx context.Context, stub APICall, callSettings CallSetting
 	backoffSettings := callSettings.RetrySettings.BackoffSettings
 	delay := backoffSettings.DelayTimeoutSettings.Initial
 	timeout := backoffSettings.RPCTimeoutSettings.Initial
+
 	for {
-		// If the deadline is exceeded...
-		if ctx.Err() != nil {
-			return ctx.Err()
-		}
-		timeoutCtx, _ := context.WithTimeout(ctx, backoffSettings.RPCTimeoutSettings.Max)
-		timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
+		timeoutCtx, _ := context.WithTimeout(ctx, timeout)
 		err := stub(timeoutCtx)
 		code := grpc.Code(err)
 		if code == codes.OK {
 			return nil
 		}
 		if !retrySettings.RetryCodes[code] {
-			return err
+			return invokeError{grpcErr: err}
 		}
-		delayCtx, _ := context.WithTimeout(ctx, backoffSettings.DelayTimeoutSettings.Max)
-		delayCtx, _ = context.WithTimeout(delayCtx, delay)
-		<-delayCtx.Done()
 
-		delay = scaleDuration(delay, backoffSettings.DelayTimeoutSettings.Multiplier)
-		timeout = scaleDuration(timeout, backoffSettings.RPCTimeoutSettings.Multiplier)
+		select {
+		case <-ctx.Done():
+			return invokeError{ctxErr: ctx.Err(), grpcErr: err}
+		case <-time.After(delay):
+		}
+
+		delay = scaleDuration(delay, backoffSettings.DelayTimeoutSettings)
+		timeout = scaleDuration(timeout, backoffSettings.RPCTimeoutSettings)
 	}
 }
 
 // Invoke calls stub with a child of context modified by the specified options.
+// If the returned error is not nil, it will be an InvokeError.
 func Invoke(ctx context.Context, stub APICall, opts ...CallOption) error {
 	settings := &CallSettings{}
 	callOptions(opts).Resolve(settings)
 	ctx = ensureTimeout(ctx, settings.Timeout)
 	if len(settings.RetrySettings.RetryCodes) > 0 {
 		return invokeWithRetry(ctx, stub, *settings)
 	}
-	return stub(ctx)
+	if err := stub(ctx); err != nil {
+		return invokeError{grpcErr: err}
+	}
+	return nil
+}
+
+// InvokeError records the GRPC error from the last completed GRPC call.
```

Comment:
Do you need an interface type here? Can the function just return a concrete &InvokeError (where InvokeError is a struct)?